### ISSUE TITLE
Twitter: Get around rate limits for testing

### ISSUE THREAD

**Issue by ccarey** [Data4Democracy/collect-social#35]
The Twitter REST API has some pretty strict [rate limits](https://dev.twitter.com/rest/public/rate-limiting) for requests. This is making testing very difficult for us! We need a way to bypass/circumvent these limits. 

One suggestion was to use [vcr](https://vcrpy.readthedocs.io/en/latest/), which has also been used for [testing tweepy](https://github.com/tweepy/tweepy/tree/97344add2ffbe78e60e47bc29af40bb6d909ee51/tests).

**Comment by arnold-jr** [comment]
I think it's probably best to go for the [Streaming source](url) rather than the REST API. This [StackOverflow post](http://stackoverflow.com/questions/34962677/twitter-streaming-api-limits), indicates that streaming rates will only be throttled when they exceed a threshold. There are twitter-specific libraries (e.g. [Python Twitter Tools](https://github.com/sixohsix/twitter/blob/master/twitter/stream_example.py)) that already have this implemented.

Still, streaming is only a subset of all tweets, and only one connection can be made per account. Furthermore, using multiple accounts is fruitless because they will see the same subset of all tweets. So if collecting all of them is necessary, then enterprise Twitter via [GNIP](https://gnip.com/) seems to be basically the only option

**Comment by bstarling** [comment]
Hey @arnold-jr thanks or dropping in to help with the project. If you aren't already there, we invite you to also join us in our D4D slack channel.

We do have multiple implementations of twitter streaming. The issue with streaming in this context is you have to already be following a person/topic in order to get the tweet (as you mentioned even then it's not all tweets). 

Once you have identified a particular account that is of interest, you need a way to go back and gather as historic tweets as possible (Given API limits). The issue @ccarey has raised is in regards to testing this "ad hoc" gathering. We need a set of previously gathered tweets to run through our tests to avoid hitting the live API when doing regression/functional testing on new code. I.E. a way to mock the twitter API so we're not running into rate limits while testing.

RE: GNIP it has come up but we're trying to make as much progress as we can without.

**Comment by alejandrox1** [comment]
@bstarling So all we are looking for here is to gather all tweets for a given user?

**Comment by bstarling** [comment]
This issue the outcome of trying to test the code to gather all tweets for a given user or all replies to a given tweet. The actual process gathering the tweets is a different issue.

This issue is specific to our need for a better way to test our code rather than hitting the live API each change. As @ccarey mentioned it looks like tweepy uses vcr to record http responses and replay them as part of their test suite.

**Comment by bstarling** [comment]
I have been working on this in the `twitter-refactor` branch https://github.com/Data4Democracy/collect-social/blob/twitter-refactor/tests/test_twitter_api.py

**Comment by domingohui** [comment]
If we are concerned about limits on api requests, [Loklak](https://github.com/loklak/loklak_python_api) may help!
Take a look at [a simple query](http://loklak.org/api/search.json?timezoneOffset=300&q=from%3ArealDonaldTrump)
TL;DR
It's basically a P2P scraper a social media platform. It's anonymous too, so no api keys needed :) No limits.
I would love to dig deeper and potentially use it here.

EDIT: One downside is that Loklak may not support all social media platforms. We would need to use the default api as a fallback in that case.
